### PR TITLE
Fix Training sort

### PR DESCRIPTION
**Pull request by samhjn**

### CODE REVIEW

File: vj4/util/tools.py [hunk 1]
```diff
@@ -18,5 +19,9 @@
         await module.ensure_indexes()
 
 
+async def dedupe(list):
+  return reduce(lambda x, y: x if y in x else x + [y], [[], ] + list)
```
Comment (twd2) [inline vj4/util/tools.py]: poor performance :(

File: .travis.yml [hunk 2]
```diff
@@ -10,6 +10,7 @@ python:
 install:
 - pip install -r requirements.txt
 - node scripts/fix_abroad_shrinkwrap.js
+- npm config set registry https://registry.npm.taobao.org
```
Comment (breeswish) [inline .travis.yml]: You can just remove "node scripts/fix_abroad_shrinkwrap.js"

File: vj4/handler/training.py [hunk 3]
```diff
@@ -13,6 +13,7 @@
 from vj4.handler import base
 from vj4.util import json
 from vj4.util import pagination
+from vj4.util.misc import dedupe
```
Comment (twd2) [inline vj4/handler/training.py]: should import `misc`

File: vj4/handler/training.py [hunk 4]
```diff
@@ -29,8 +30,8 @@ def _parse_dag_json(dag):
         raise error.ValidationError('dag')
       new_node = {'_id': int(node['_id']),
                   'title': str(node.get('title', '')),
-                  'require_nids': list(set(map(int, node['require_nids']))),
-                  'pids': list(set(map(document.convert_doc_id, node['pids'])))}
+                  'require_nids': dedupe(map(int, node['require_nids'])),
```
Comment (twd2) [inline vj4/handler/training.py]: `misc.dedup`

File: vj4/util/misc.py [hunk 5]
```diff
@@ -94,3 +94,14 @@ def format_seconds(seconds):
 def base64_encode(str):
   encoded = base64.b64encode(str.encode())
   return encoded.decode()
+
+
+def dedupe(list):
```
Comment (twd2) [inline vj4/util/misc.py]: `dedup`

File: vj4/util/tools.py [hunk 6]
```diff
@@ -1,6 +1,7 @@
 import logging
 import pkgutil
 from os import path
+from functools import reduce
```
Comment (twd2) [inline vj4/util/tools.py]: ???

File: vj4/util/misc.py (hunk 5 above)
Comment (samhjn) [reply vj4/util/misc.py]: dedupe is okay

File: .travis.yml (hunk 2 above)
Comment (breeswish) [reply .travis.yml]: I mean, by removing "node scripts/fix_abroad_shrinkwrap.js", you will get a shrinkwrap of npm.taobao.org, and you don't need to run `npm config set` any more.

**Pull request merged**